New Project > Select Repository

AlexTugarev opened the pull request:
### This PR introduces the project creation flow (gitpod.io/new)

---

### GH App Change

⚠️ Please note, this change requires a significant GH App config change to be aligned with any deployment: instead of redirecting to an URL of the dashboard, it should redirect to the `/api/apps/github/setup` route introduced here. 

<img width="772" alt="Screen Shot 2021-06-14 at 05 39 29" src="https://user-images.githubusercontent.com/914497/121836058-0640ae00-ccd3-11eb-9877-fd8c8001426f.png">

---

### Preview environment
The URL is https://at-select-repo.staging.gitpod-dev.com/new

⚠️ Please note, this preview environment requires an additional re-deployment with werft from within a Gitpod workspace using `werft run github -s chart/values.yaml -f`, where the provided values should contain the GH App config. In addition, a secret with the private key of the GH App should be created like this `kubectl create secret generic server-github-app-cert ...`

---
# New Project

The drop-down list contains the user's account by default. If the GH App is installed on Orgs of the users and the user granted `read:org` permissions, these accounts should be visible as well.

<img width="403" alt="Screen Shot 2021-06-14 at 05 52 26" src="https://user-images.githubusercontent.com/914497/121836911-bcf15e00-ccd4-11eb-9bfa-e27310d5d87b.png">


<img width="1100" alt="Screen Shot 2021-06-14 at 05 29 35" src="https://user-images.githubusercontent.com/914497/121835509-c200de00-ccd1-11eb-87eb-f8a1e52ba5b8.png">

- [ ] This list should be filtered to exclude repositories already in used in any project. 

## Select Git Provider

<img width="1113" alt="Screen Shot 2021-06-14 at 05 29 41" src="https://user-images.githubusercontent.com/914497/121835530-cdeca000-ccd1-11eb-9634-f09625bd73c1.png">

## Reconfigure GH App installations

You will see the GH App configuration in a pop-up window. For the preview environment this is a test app of mine.

<img width="676" alt="Screen Shot 2021-06-14 at 05 30 05" src="https://user-images.githubusercontent.com/914497/121835543-da70f880-ccd1-11eb-86c3-3e3cbc5245d9.png">

# Select a team

Listing of all teams of a user. 

- [x] Incorporate "New Team" inline

<img width="1105" alt="Screen Shot 2021-06-14 at 05 30 18" src="https://user-images.githubusercontent.com/914497/121835571-e8bf1480-ccd1-11eb-9beb-0a0ea9c9631d.png">

# List projects of a team

This is a starting point for the Projects overview. 

<img width="1125" alt="Screen Shot 2021-06-14 at 05 30 52" src="https://user-images.githubusercontent.com/914497/121835590-f70d3080-ccd1-11eb-903f-47f846fc2cac.png">

Closes https://github.com/gitpod-io/gitpod/issues/4352

AlexTugarev commented on components/server/ee/src/prebuilds/github-app.ts:
```diff
@@ -82,6 +80,12 @@ export class GithubApp {
             res.redirect(301, this.getBadgeImageURL());
         });
 
+        options.getRouter && options.getRouter('/whoami').get('/', async (req: express.Request, res: express.Response, next: express.NextFunction) => {
```
todo: remove

AlexTugarev commented on components/server/src/workspace/gitpod-server-impl.ts:
```diff
@@ -1810,4 +1809,27 @@ export class GitpodServerImpl<Client extends GitpodClient, Server extends Gitpod
+
+    //#region Projects concerns
+    //
+    async getProviderRepositoriesForUser(params: { provider: string }): Promise<ProviderRepository[]> {
```
this is implemented in EE space because the github-app.ts is also there. 

@svenefftinge, it's ok to move the github-app.ts to the non-EE sources?

svenefftinge replied:
Yes, that is fine

AlexTugarev commented on components/server/ee/src/prebuilds/github-app.ts:
```diff
@@ -103,6 +107,23 @@ export class GithubApp {
+            const state = req.query.state;
+            res.redirect(`https://github.com/apps/${slug}/installations/new?state=${state}`)
+        });
+        options.getRouter && options.getRouter('/setup').get('/', async (req: express.Request, res: express.Response, next: express.NextFunction) => {
```
todo: re-introduce `req.query.state` which will be passed on `/reconfigure` in order to identify the origins of this request. IF we get called here without `state`, we can assume, that this is call from the marketplace, thus redirecting to `/install-github-app` should be done.

gtsiolis commented on components/dashboard/src/projects/NewProject.tsx:
```diff
@@ -0,0 +1,314 @@
+                        <select name="selectAccount" value={selectedAccount} className="w-full"
+                            onChange={(e) => onSelectAccount(e.target.value)}>
+                            {accounts.map(a => (<option key={`account-${a}`} value={a}>{`${a} (${provider})`}</option>))}
+                            {isGitHub() && (<option value="addOrg">Add GH Org</option>)}
```
**nitpick**: Reminder to avoid acronyms like GH here. 😇

AlexTugarev replied:
Very well. Just realized how surprising the comments are, while changing code at the very same time ;-)

gtsiolis commented on components/dashboard/src/projects/NewProject.tsx:
```diff
@@ -0,0 +1,314 @@
+                    <div className="text-center text-gray-500">
+                        Select a Git provider first and continue with your repositories.
+                    </div>
+                    <div className="mt-6 flex flex-col space-y-3 items-center">
```
**nitpick**: A minor spacing suggestion.

```suggestion
                    <div className="mt-6 flex flex-col space-y-3 items-center pb-8">
```

gtsiolis commented on components/dashboard/src/projects/NewProject.tsx:
```diff
@@ -0,0 +1,314 @@
+                    </div>
+                    {noOrgs && (
+                        <div className="text-gray-500 mx-auto text-center">
+                            Missing organizations? <a href="javascript:void(0)" onClick={e => grantReadOrgPermissions()} className="text-gray-400 underline underline-thickness-thin underline-offset-small hover:text-gray-600">Grant permissions</a>
```
**issue**: This is not applicable when switching to a GitLab provider, right?

AlexTugarev replied:
At the moment this is GitHub specific. 
TODO: disable for GitLab

gtsiolis reviewed (commented):
Great to see this coming to life @AlexTugarev! Left some minor UX comments below. 🏀

AlexTugarev commented on components/dashboard/src/projects/NewProject.tsx:
```diff
@@ -0,0 +1,314 @@
+                </div>
+                <div className="p-3 bg-gray-100">
+                    <div className="text-gray-500 text-center">
+                        Repository not found? <a href="javascript:void(0)" onClick={e => reconfigure()} className="text-gray-400 underline underline-thickness-thin underline-offset-small hover:text-gray-600">Reconfigure</a>
```
Yes. 

When selecting `Reconfigure` and editing the installations of the App on GH, you may have either: 
1. Install on new Account/Org
2. Selected additional Repos (or selected all)
3. Deleted the installation from an Account/Org

All these kinds of changes needs to be reflected in the action (Reconfigure) and also updated after the action is finished, which I doubt happens in all cases. At least after deletion, we're not getting redirected :-( So in that case, where the user does (need to) close the GH pop-up window, we need to update!

AlexTugarev commented on components/dashboard/src/projects/NewProject.tsx:
```diff
@@ -0,0 +1,314 @@
+                                        </div>
+                                        <div className="w-4/12 flex justify-end">
+                                            <div className="flex self-center hover:bg-gray-200 dark:hover:bg-gray-700 rounded-md cursor-pointer opacity-0 group-hover:opacity-100">
+                                                <button className="primary py-1" onClick={() => selectRepo(r)}>Select</button>
```
would that apply to the `Select Team` page as well?

gtsiolis replied:
In case the project is already part of a team we could trigger a modal with a CTA button to request access to the team, instead of having conditional buttons here _Add Project_ and _Select Team_. WDYT?

AlexTugarev commented:
https://github.com/gitpod-io/gitpod/pull/4443#discussion_r650974879

> In case the project is already part of a team we could trigger a modal with a CTA button to request access to the team, instead of having conditional buttons here Add Project and Select Team. WDYT?

@gtsiolis, thanks for bringing this up! It's a rather conceptual question, so let's discuss. 

If a repository (aka project) is already linked to a team, what is the implication to the overall concept and the creation flow?Potentially incomplete list of options:
1. Exclude taken repositories on Create Project page.
2. Disable taken repositories on Create Project page.
3. Allow every accessible repo on Create Project page, then add an invitation method.

svenefftinge commented:
/werft run

AlexTugarev commented:
The check for the state parameter on setup callbacks provides a nice migration path. When installing the app from GH, we get redirected to `/install-github-app` like before

<img width="1661" alt="Screen Shot 2021-06-15 at 14 13 11" src="https://user-images.githubusercontent.com/914497/122050732-0e3c4300-cde4-11eb-99d0-292c3c75dd4a.png">
<img width="1657" alt="Screen Shot 2021-06-15 at 14 13 22" src="https://user-images.githubusercontent.com/914497/122050738-10060680-cde4-11eb-9797-37cfb8f8d75f.png">

svenefftinge commented:
- [ ] When I add an org, it replaces the org I previously added.

svenefftinge commented:
- [ ] The new project flow should not have any menus or so. See design docs.

AlexTugarev commented:
> When I have added the 'orgs' scope, I still don't see the orgs for which installations exist

@svenefftinge, that is true as long as the gitpod-dev OAuth app isn't approved for the org. On https://at-select-repo.staging.gitpod-dev.com/integrations you can select `Manage github.com` or directly navigate to 
 https://github.com/settings/connections/applications/51329bab66290acdf39f and make sure you can approve an organization to test with.

AlexTugarev commented:
@svenefftinge, I updated the PR with following changes:

- [ ] When I add an org, it replaces the org I previously added.

I'm not sure, what that means. The `Reconfigure` action (showing GitHub's installation page) combines several options: installing, updating, or deleting. 

- [x] Even when I already have a project for an org, I don't see that org anymore in the list.

That should be solved now. `knownOrgs` are preserved in `User.additionalData`, so that when you come back later to `/new`, you should be able to browse the repos of that org. 

- [x] When I land on new, nothing in the combo box is selected.

That might be just related to the missing org, I suppose. In general it takes up to seconds if you have many orgs accessible to the OAuth App of the Gitpod installation. 

- [x] When I added an app installation, the newly added installation is not selected

Done. They should be selected now properly. Also adding `updateAt` for installations and repositories. Repositories are sorted now. 

- [x] When I have added the 'orgs' scope, I still don't see the orgs for which installations exist

This is fixed in the sense that known orgs are preserved on backend now.

AlexTugarev commented:
> The new project flow should not have any menus or so. See design docs. 

@svenefftinge and @gtsiolis, I aligned with /new-teams on that. 

I'm happy to reduce the menus. For both?

svenefftinge commented on components/gitpod-db/src/typeorm/migration/1622468446118-TeamsAndProjects.ts:
```diff
@@ -11,7 +11,7 @@ export class TeamsAndProjects1622468446118 implements MigrationInterface {
         await queryRunner.query("CREATE TABLE IF NOT EXISTS `d_b_team` (`id` char(36) NOT NULL, `name` varchar(255) NOT NULL, `slug` varchar(255) NOT NULL, `creationTime` varchar(255) NOT NULL, `deleted` tinyint(4) NOT NULL DEFAULT '0', `_lastModified` timestamp(6) NOT NULL DEFAULT CURRENT_TIMESTAMP(6) ON UPDATE CURRENT_TIMESTAMP(6), PRIMARY KEY (`id`), KEY `ind_dbsync` (`_lastModified`)) ENGINE= ...
         await queryRunner.query("CREATE TABLE IF NOT EXISTS `d_b_team_membership` (`id` char(36) NOT NULL, `teamId` char(36) NOT NULL, `userId` char(36) NOT NULL, `role` varchar(255) NOT NULL, `creationTime` varchar(255) NOT NULL, `deleted` tinyint(4) NOT NULL DEFAULT '0', `_lastModified` timestamp(6) NOT NULL DEFAULT CURRENT_TIMESTAMP(6) ON UPDATE CURRENT_TIMESTAMP(6), PRIMARY KEY (`id`), KEY `i ...
-        await queryRunner.query("CREATE TABLE IF NOT EXISTS `d_b_project` (`id` char(36) NOT NULL, `cloneUrl` varchar(255) NOT NULL, `teamId` char(36) NOT NULL, `appInstallationId` varchar(255) NOT NULL, `creationTime` varchar(255) NOT NULL, `deleted` tinyint(4) NOT NULL DEFAULT '0', `_lastModified` timestamp(6) NOT NULL DEFAULT CURRENT_TIMESTAMP(6) ON UPDATE CURRENT_TIMESTAMP(6), PRIMARY KEY (`i ...
+        await queryRunner.query("CREATE TABLE IF NOT EXISTS `d_b_project` (`id` char(36) NOT NULL, `name` varchar(255) NOT NULL, `cloneUrl` varchar(255) NOT NULL, `teamId` char(36) NOT NULL, `appInstallationId` varchar(255) NOT NULL, `creationTime` varchar(255) NOT NULL, `deleted` tinyint(4) NOT NULL DEFAULT '0', `_lastModified` timestamp(6) NOT NULL DEFAULT CURRENT_TIMESTAMP(6) ON UPDATE CURRENT ...
```
do you plan to manually update the staging DB?

AlexTugarev replied:
good point! this is already on staging. 
yes, it would be best to do a manual update.

svenefftinge commented on components/gitpod-protocol/src/protocol.ts:
```diff
@@ -1205,6 +1208,27 @@ export interface Project {
     deleted?: boolean;
 }
 
+export interface ProjectInfo {
```
Why copying the whole `Project` shape instead of referencing it?

AlexTugarev replied:
Just learned that `DBProject` isn't implementing `Project` at all.
So, yeah it doesn't make sense to duplicate. 

Besides that, we introduced `xInfo` shapes to augment entities, to make it works nicer with the frontend.

svenefftinge commented on components/server/ee/src/workspace/gitpod-server-impl.ts:
```diff
@@ -1421,4 +1427,164 @@ export class GitpodServerEEImpl extends GitpodServerImpl<GitpodClient, GitpodSer
+    //#region Projects concerns
+    //
+    async getProviderRepositoriesForUser(params: { provider: string, hints?: object }): Promise<ProviderRepository[]> {
+        const user = this.checkAndBlockUser("getProviderRepositoriesForUser");
```
this method should go the respective GitProvider impl. We should not add dependencies to GitHub (octokit) across our code.

AlexTugarev replied:
+1 for moving out. I'm thinking of a `GitHubAppSupport` injectable.  

FYI we have two octokit dependencies
- first in our GitHub support (auth provider, context parsers, etc.)
- second comes in via probot
Both parts live in different contexts. More specifically: the GitHub support is instantiated per Git Provider host. 

BTW it's ~impossible~ impractical to sync both octokit versions both. Also, obtaining GH App's authenticated API (octokit) is done via `GitHubApp.server.probotApp.auth()`.

svenefftinge reviewed (approved):
Works well. It needs additional passes, of course, but since it is not activated / visible by default it would be good to merge ASAP. So we can iterate in smaller batches.

jankeromnes commented on components/gitpod-db/src/typeorm/project-db-impl.ts:
```diff
@@ -27,4 +28,19 @@ export class ProjectDBImpl implements ProjectDB {
+
+        const project: Project = {
+            id: uuidv4(),
+            name,
```
Nit: Since `name` is also used as a `slug`, maybe validate it here against a regex? For example like so:

https://github.com/gitpod-io/gitpod/blob/903c8545e5eba0365ba008465ac634b2cfd0c254/components/gitpod-db/src/typeorm/team-db-impl.ts#L74-L76
